docs: added information on BrowserTestInfo (#2057)
vaadin/docs: articles/testing/end-to-end/advanced-concepts.adoc

---
title: Advanced Testing Methods 
order: 80
---

= Advanced Testing Methods

There are a few advanced testing methods you may want to consider: waiting for Vaadin; waiting until a particular condition is met; scrolling; profiling test execution time; and running tests in parallel.

These testing methods aren't typically needed. For example, situations in which you might need to disable automatic waiting or scrolling in a view are rare. In such cases, you've probably encountered a bug in the software. Nevertheless, these testing methods are explained here for when they are needed.

== Waiting for Vaadin

> Web pages load typically when immediately rendered by the browser. In such applications, you can test the page elements immediately after the page is loaded. In Vaadin and other Single-Page Applications (SPAs), rendering is done by JavaScript code, asynchronously. Therefore, you need to wait until the server has given its response to an AJAX request and the JavaScript code finishes rendering the UI.

Reviewer: ```suggestion
Web pages are typically loaded and rendered immediately by the browser. In such applications, you can test the page elements immediately after the page is loaded. In Vaadin and other Single-Page Applications (SPAs), rendering is done by JavaScript code, asynchronously. Therefore, you need to wait until the server has given its response to an AJAX request and the JavaScript code finishes rendering the UI.
```
This was a bit confusing
